20221012-add-type-restrictions-to-dsl-syntax.md

```diff
@@ -0,0 +1,117 @@
+- [What is a user?](https://openfga.dev/docs/concepts#what-is-a-user)
+- Reverse Expansion: Normally [expand](https://openfga.dev/docs/interacting/relationship-queries#expand) takes an object and relation and returns the first leaves of users who are related. Reverse expand would be the opposite, taking a user and a relation and returning the objects which are related.
+
+## Motivation
+
+### Improve the developer experience
+
+Adding type restrictions improves the understandability of authorization models. One may see immediately that certain relations may only have certain types of users. For example, parents of folders can only be of type folder, or readers of documents can only be of type user.
+
+### Optimize ListObjects
+
+Type restrictions will allow us to optimize the [ListObjects](https://github.com/openfga/rfcs/blob/main/20220714-listObjects-api.md) endpoint. ListObjects needs to traverse the permission graph in reverse, and by restricting the types of users that may be related to objects, we are able to reduce the number of edges we must traverse from each node.
+
+## Type restrictions
+
+### Types with no relations
+
+It now becomes useful to allow types with no relations. For example, we may want to specify a `user` type which is the type for all concrete users. This will look like
+```
+type user
+```
+
+### Type restrictions
+
+Let A be an authorization model with set of types T and set of relations R. A type restriction on A can be seen as a set-valued function from TxR to TxR*, where R* is R with the addition of an "empty" relation. That is, a type restriction takes any (type, relation) in A and takes it to the set of type-relation pairs, with possibly empty relation, that it may be directly related to.
+
+#### Examples
+
+1. Suppose A is an authorization model with a type folder and relation parent on folder. We may have a type restriction on A that takes (folder, parent) to the set {(folder, \_)}, meaning that parents of folders must be folders. (The "_" is used to indicate the empty relation.)
+ 2. Suppose A is an authorization model with type document and relation reader on document. We may have a type restriction on A that takes (document, reader) to the set {(user, \_), (group, member)}, meaning that the only users that may be readers of a document are users with type user, or a userset of group members.
+
+
+### Syntax
+
+Let A be an authorization model with set of types T and set of relations R, and recall that R* is the set R with the addition of an "empty" relation. We will write an element (t, r) of TxR* as t#r. The special case (t, _) is simply written as t. If the type restriction takes (t, r) to a set {(t1, r1), (t2, r2)}, then in our model we will write this as
+```
+type t
+    relations
+        define r: [t1#r1, t2#r2] as ...
+```
+If the type restriction takes (t, r) to the empty set we will simply write nothing in our model.
```

Comment:
However, for empty set, there will be no entry point for empty set.  In this case, should we disallow as the relation cannot be used with no entry point?